core/sql/lib_mgmt/src/main/java/org/trafodion/libmgmt/SyncLibUDF.java

```diff
@@ -0,0 +1,170 @@
+*
+*  http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*
+* @@@ END COPYRIGHT @@@
+ */
+package org.trafodion.libmgmt;
+
+import java.io.File;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import org.trafodion.sql.udr.UDR;
+import org.trafodion.sql.udr.UDRInvocationInfo;
+import org.trafodion.sql.udr.UDRPlanInfo;
+import org.trafodion.sql.udr.UDRException;
+
+public class SyncLibUDF extends UDR {
+
+    // default constructor
+    public SyncLibUDF()
+    {}
+
+    @Override
+    public void describeParamsAndColumns(UDRInvocationInfo info)
+        throws UDRException
+    {
+        // this TMUDF takes no table-valued inputs, two string
+        // parameters, and generates a table with a single integer
+        // value. It assumes that it was created with the following
+        // DDL, without specifying parameters or return values:
+        //
+        //  create table_mapping function SyncLibUDF()
+        //  external name 'org.trafodion.libmgmt.SyncLibUDF'
+        //  library ...;
+        //
+        // The call is like this: select ... from udf(synclibudf(<op>,<file>))
+        //
+        // <op> is 'c' to create/copy a file from the HDFS staging area for
+        //             the current user into a local file on every node, or
+        //         'd' to delete/drop a local file on every node
+        // <file> is an unqualified file name (no '/' in the name)
+        //
+        // Example:
+        //
+        // select * from udf("_LIBMGR_".synclibudf('c', 'mylib.jar'));
+        //
+        // INSTANCE_NUM
+        // ------------
+        //
+        //            0
+        //            1
+        //
+        // --- 2 row(s) selected.
+        // >>
+
+
+        if (info.par().getNumColumns() != 2)
+            throw new UDRException(38970, "This UDF needs to be called with two input parameters");
+        if (info.getNumTableInputs() > 0)
+            throw new UDRException(38971, "This UDF needs to be called without table-valued inputs");
+
+        for (int i=0; i<2; i++)
+            info.addFormalParameter(info.par().getColumn(i));
+        info.out().addIntColumn("INSTANCE_NUM", false);
+    }
+
+    @Override
+    public void describeDesiredDegreeOfParallelism(UDRInvocationInfo info,
+                                                   UDRPlanInfo plan)
+        throws UDRException
+    {
+        boolean usesVirtualNodes = false;
+
+        // check for configurations with virtual nodes. Run the UDF serially
+        // in those cases, since all the virtual nodes share the same node.
```

Comment:
This is an interesting bit of smarts. Is it universal among our system UDFs? I wonder too about our optimizer: does it reduce DoP when we run on virtual nodes?